tests/providers/rushfiles/test_provider.py

```diff
@@ -245,6 +245,109 @@ async def test_validate_v1_path_root(self, provider):
+        url_body = json.dumps({
+                        'TransmitId': provider._generate_uuid(),
+                        'ClientJournalEventType': 1,
+                        'DeviceId': 'waterbutler'
+                    })
+
+        aiohttpretty.register_uri('DELETE', url, body=url_body, status=200)
+
+        await provider.delete(path)
+
+        assert aiohttpretty.has_call(method='DELETE', uri=url)
+
+    @pytest.mark.asyncio
+    async def test_delete_path_does_not_exist(self, provider):
+        path = RushFilesPath('/Gone')
+
+        with pytest.raises(exceptions.NotFoundError) as e:
+            await provider.delete(path)
+
+        assert e.value.code == 404
+        assert str(path) in e.value.message
+
+    @pytest.mark.asyncio
+    @pytest.mark.aiohttpretty
+    async def test_delete_root(self, provider):
+        path = RushFilesPath('/', _ids=[provider.share['id']], folder=True)
+
+        with pytest.raises(exceptions.DeleteError) as e:
+            await provider.delete(path)
+
+        assert e.value.message == 'root cannot be deleted'
+        assert e.value.code == 400
+
+class TestDownload:
+
+    @pytest.mark.asyncio
+    @pytest.mark.aiohttpretty
+    async def test_download(self, provider, root_provider_fixtures):
+        body = b'dearly-beloved'
+        path = WaterButlerPath('/Tasks.xlsx',_ids=(provider.share['id'],'0f04f33f715a4d5890307f114bf24e9c'))
+        metadata = root_provider_fixtures['file_metadata_resp']
+        
+        metadata_url = provider._build_clientgateway_url(str(provider.share['id']), 'virtualfiles', path.identifier)
+        aiohttpretty.register_uri('GET', metadata_url, body=json.dumps(metadata))
+
+        url = provider._build_filecache_url(str(provider.share['id']), 'virtualfiles', metadata['Data']['UploadName'])
+        aiohttpretty.register_uri('GET', url, body=body)
+
+        result = await provider.download(path)
+        content = await result.read()
+
+        assert content == body
+
+    @pytest.mark.asyncio
+    @pytest.mark.aiohttpretty
+    async def test_download_path_is_dir(self, provider):
+        path = WaterButlerPath('/lets-go-dir/')
+        with pytest.raises(exceptions.DownloadError):
+            await provider.download(path)
+
+    @pytest.mark.asyncio
+    @pytest.mark.aiohttpretty
+    async def test_download_id_not_found(self, provider):
+        path = WaterButlerPath('/lets-go-crazy')
+        with pytest.raises(exceptions.DownloadError):
+            await provider.download(path)
+
+    @pytest.mark.asyncio
+    @pytest.mark.aiohttpretty
+    async def test_download_not_found(self, provider, root_provider_fixtures):
+        path = WaterButlerPath('/lets-go-crazy')
+        metadata = root_provider_fixtures['file_metadata_resp']
+        url = provider.build_url(str(provider.share['id']), 'virtualfiles', metadata['Data']['UploadName'])
+        aiohttpretty.register_uri('GET', url, status=404)
```

Comment:
- Wrong URL
- 404 you get when share does not exists. When file doesn't exists you get 400.
- A request won't be made either way because there is no _ids in path so this test is the same as the above